File: scripts/install-skills.test.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require "fileutils"
require "minitest/autorun"
require_relative "test_helper"

class InstallSkillsTest < Minitest::Test
  include ScriptTestRepo

  def test_force_skips_target_that_is_source
    build_script_repo("install-skills", skills: { "sample" => good_frontmatter("sample") }) do |dir|
      stdout, stderr, status = run_script(
        dir, "install-skills", "--target", File.join(dir, "skills"), "--force", "sample"
      )

      assert status.success?, stderr
      assert_includes stderr, "target is source, skipping"
      assert_empty stdout
      skill_md = File.join(dir, "skills", "sample", "SKILL.md")
      assert File.exist?(skill_md), "expected #{skill_md} to survive"
    end
  end

  def test_force_copy_replaces_existing_symlink
    build_script_repo("install-skills", skills: { "sample" => good_frontmatter("sample") }) do |dir|
      target_dir = File.join(dir, "target")
      FileUtils.mkdir_p(target_dir)
      FileUtils.ln_s(File.join(dir, "skills", "sample"), File.join(target_dir, "sample"))

      stdout, stderr, status = run_script(
        dir, "install-skills", "--target", target_dir, "--force", "--mode", "copy", "sample"
      )

      assert status.success?, stderr
      assert_includes stdout, "copy sample"
      refute File.symlink?(File.join(target_dir, "sample"))
      copied = File.join(target_dir, "sample", "SKILL.md")
      source = File.join(dir, "skills", "sample", "SKILL.md")
      assert File.exist?(copied), "expected copy at #{copied}"
      assert File.exist?(source), "expected source at #{source} to remain"
    end
  end

  def test_list_prints_available_skills_sorted
    skills = %w[charlie alpha bravo].to_h { |name| [name, good_frontmatter(name)] }
    build_script_repo("install-skills", skills: skills) do |dir|
      stdout, _stderr, status = run_script(dir, "install-skills", "--list")

      assert status.success?
      assert_equal %w[alpha bravo charlie], stdout.split("\n")
    end
  end

  def test_unknown_skill_fails
    build_script_repo("install-skills", skills: { "sample" => good_frontmatter("sample") }) do |dir|
      stdout, stderr, status = run_script(dir, "install-skills", "nope")

      refute status.success?
      assert_includes stderr, "unknown skill(s): nope"
      assert_empty stdout
    end
  end

  def test_dry_run_leaves_target_untouched
    build_script_repo("install-skills", skills: { "sample" => good_frontmatter("sample") }) do |dir|
      target_dir = File.join(dir, "target")
      stdout, _stderr, status = run_script(dir, "install-skills", "--target", target_dir, "--dry-run", "sample")

      assert status.success?
      assert_includes stdout, "would symlink sample"
      refute File.exist?(target_dir), "dry-run must not create the target directory"
    end
  end

  def test_default_skip_if_exists
    build_script_repo("install-skills", skills: { "sample" => good_frontmatter("sample") }) do |dir|
      target_dir = File.join(dir, "target")
      FileUtils.mkdir_p(File.join(target_dir, "sample"))
      original_marker = File.join(target_dir, "sample", "marker")
      File.write(original_marker, "keep me")

      stdout, stderr, status = run_script(dir, "install-skills", "--target", target_dir, "sample")

      assert status.success?, stderr
      assert_includes stderr, "exists, skipping"
      assert_empty stdout
      assert File.exist?(original_marker), "non-forced reinstall must not touch an existing target"
    end
  end

  def test_invalid_mode_fails
    build_script_repo("install-skills", skills: { "sample" => good_frontmatter("sample") }) do |dir|
      stdout, stderr, status = run_script(dir, "install-skills", "--mode", "bogus", "sample")

      refute status.success?
      assert_equal 2, status.exitstatus
      assert_includes stderr, "invalid --mode: bogus"
      assert_empty stdout
    end
  end
end

File: scripts/install-skills.test.py
from __future__ import annotations

import os
import shutil
import subprocess
import sys
import tempfile
import unittest
from pathlib import Path


SCRIPT = Path(__file__).with_name("install-skills")


class InstallSkillsTest(unittest.TestCase):
    def make_repo(self, root: Path, skills: tuple[str, ...] = ("sample",)) -> Path:
        repo = root / "repo"
        (repo / "scripts").mkdir(parents=True)
        shutil.copy2(SCRIPT, repo / "scripts" / "install-skills")
        for name in skills:
            skill_dir = repo / "skills" / name
            skill_dir.mkdir(parents=True)
            (skill_dir / "SKILL.md").write_text(
                f"---\nname: {name}\ndescription: {name}\n---\n", encoding="utf-8"
            )
        return repo

    def run_installer(self, repo: Path, *args: str) -> subprocess.CompletedProcess[str]:
        return subprocess.run(
            [sys.executable, str(repo / "scripts" / "install-skills"), *args],
            check=False,
            capture_output=True,
            text=True,
        )

    def test_force_skips_target_that_is_source(self) -> None:
        with tempfile.TemporaryDirectory() as tmp:
            repo = self.make_repo(Path(tmp))
            result = self.run_installer(
                repo, "--target", str(repo / "skills"), "--force", "sample"
            )

            self.assertEqual(result.returncode, 0, result.stderr)
            self.assertIn("target is source, skipping", result.stderr)
            self.assertEqual(result.stdout, "")
            self.assertTrue((repo / "skills" / "sample" / "SKILL.md").exists())

    def test_force_copy_replaces_existing_symlink(self) -> None:
        with tempfile.TemporaryDirectory() as tmp:
            root = Path(tmp)
            repo = self.make_repo(root)
            target = root / "target"
            target.mkdir()
            try:
                os.symlink(repo / "skills" / "sample", target / "sample", target_is_directory=True)
            except OSError as error:
                if os.name == "nt":
                    self.skipTest(f"symlink creation unavailable: {error}")
                raise

            result = self.run_installer(
                repo, "--target", str(target), "--force", "--mode", "copy", "sample"
            )

            self.assertEqual(result.returncode, 0, result.stderr)
            self.assertIn("copy sample", result.stdout)
            self.assertFalse((target / "sample").is_symlink())
            self.assertTrue((target / "sample" / "SKILL.md").exists())
            self.assertTrue((repo / "skills" / "sample" / "SKILL.md").exists())

    def test_list_is_sorted(self) -> None:
        with tempfile.TemporaryDirectory() as tmp:
            repo = self.make_repo(Path(tmp), ("zeta", "alpha"))
            result = self.run_installer(repo, "--list")

            self.assertEqual(result.returncode, 0, result.stderr)
            self.assertEqual(result.stdout, "alpha\nzeta\n")

    def test_dry_run_does_not_create_target(self) -> None:
        with tempfile.TemporaryDirectory() as tmp:
            root = Path(tmp)
            repo = self.make_repo(root)
            target = root / "missing" / "skills"
            result = self.run_installer(repo, "--target", str(target), "--dry-run", "sample")

            self.assertEqual(result.returncode, 0, result.stderr)
            self.assertEqual(result.stdout, f"would symlink sample -> {target / 'sample'}\n")
            self.assertFalse(target.exists())

    def test_unknown_skill_reports_available_names(self) -> None:
        with tempfile.TemporaryDirectory() as tmp:
            repo = self.make_repo(Path(tmp))
            result = self.run_installer(repo, "missing")

            self.assertEqual(result.returncode, 1)
            self.assertEqual(
                result.stderr,
                "unknown skill(s): missing\navailable: sample\n",
            )

    def test_invalid_mode_exits_two(self) -> None:
        with tempfile.TemporaryDirectory() as tmp:
            repo = self.make_repo(Path(tmp))
            result = self.run_installer(repo, "--mode", "invalid")

            self.assertEqual(result.returncode, 2)
            self.assertEqual(result.stderr, "invalid --mode: invalid\n")


if __name__ == "__main__":
    unittest.main()
